darling/src/libsystem_kernel/emulation/src/xnu_syscall/bsd/impl/misc/clonefile.c

```diff
@@ -1,13 +1,158 @@
 #include <darling/emulation/xnu_syscall/bsd/impl/misc/clonefile.h>
 
+#include <sys/errno.h>
+#include <sys/stat.h>
+#include <sys/time.h>
 #include <signal.h>
+#include <string.h>
+#include <time.h>
 
-#include <darling/emulation/conversion/duct_errno.h>
+#include <darling/emulation/common/base.h>
+#include <darling/emulation/conversion/common_at.h>
+#include <darling/emulation/conversion/errno.h>
+#include <darling/emulation/conversion/fcntl/open.h>
+#include <darling/emulation/linux_premigration/linux-syscalls/linux.h>
+#include <darling/emulation/linux_premigration/vchroot_expand.h>
+#include <darling/emulation/other/mach/lkm.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/fcntl/openat.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/close.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/fchmod.h>
+#include <darling/emulation/xnu_syscall/bsd/impl/unistd/fchown.h>
+
+#ifndef MAX_COPY_CHUNK
+#define MAX_COPY_CHUNK ((size_t)0x7ffff000)
+#endif
 
 long sys_clonefileat(int src_fd, const char* src_path, int dest_fd, const char* dest_path, uint32_t flags) {
 	return -ENOSYS;
 }
 
 long sys_fclonefileat(int src_fd, int dest_fd, const char* dest_path, uint32_t flags) {
-	return -ENOSYS;
+    struct stat st;
+
+    if (fstat(src_fd, &st) < 0) {
```

Comment:
I prefer if we stick with the direct syscall implementation instead.
```suggestion
    if (sys_fstat(src_fd, &st) < 0) {
```